File: androidApp/src/main/kotlin/com/malaria/android/services/SettingsStore.kt
package com.malaria.android.services

import com.malaria.android.data.LanguagePreference
import com.malaria.android.data.OnboardingLanguage
import com.malaria.android.data.dao.AuditDao
import com.malaria.android.data.dao.ClinicianDao
import com.malaria.android.data.entities.AuditAction
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Settings facade backed by the audit log (spec §11, §8). Mirrors the iOS
 * `SettingsStore` value-for-value.
 *
 * Clinic-level config (name, jurisdiction, lawful basis) and the inference
 * policy (threshold, default model, auto-logout) live in the audit log as
 * the source of truth — written during Phase 1 admin provisioning
 * (`clinic_configured`, `inference_policy_set`) and updated by Settings
 * edits via `threshold_changed`, `default_model_changed`,
 * `auto_logout_changed`. Holding the latest values here avoids re-parsing
 * metadata JSON on every render.
 *
 * `language` lives in [LanguagePreference] (DataStore) because it must
 * survive `Reset device` (spec §10) where the audit log gets a fresh
 * `device_reprovisioned` entry but keeps every prior provisioning entry —
 * the chain-of-custody requirement.
 *
 * `language` is a [String] (canonical lowercase) rather than the
 * `OnboardingLanguage` enum so the value matches the iOS surface field-
 * for-field — the composables convert via `OnboardingLanguage.fromCanonical`.
 *
 * `languagePreference` may be null in unit tests that want to assert just
 * the audit-driven hydration without spinning up a DataStore context.
 */
class SettingsStore(
    private val auditDao: AuditDao,
    private val audit: AuditLog,
    private val clinicians: ClinicianDao,
    private val languagePreference: LanguagePreference? = null,
) {
    private val _clinicName = MutableStateFlow<String?>(null)
    val clinicName: StateFlow<String?> = _clinicName.asStateFlow()

    private val _jurisdiction = MutableStateFlow<String?>(null)
    val jurisdiction: StateFlow<String?> = _jurisdiction.asStateFlow()

    private val _lawfulBasis = MutableStateFlow<String?>(null)
    val lawfulBasis: StateFlow<String?> = _lawfulBasis.asStateFlow()

    private val _threshold = MutableStateFlow(0.3)
    val threshold: StateFlow<Double> = _threshold.asStateFlow()

    private val _defaultModelId = MutableStateFlow("BNLeaky_Keras")
    val defaultModelId: StateFlow<String> = _defaultModelId.asStateFlow()

    private val _autoLogoutMinutes = MutableStateFlow(15)
    val autoLogoutMinutes: StateFlow<Int> = _autoLogoutMinutes.asStateFlow()

    private val _language = MutableStateFlow(OnboardingLanguage.English.canonical)
    val language: StateFlow<String> = _language.asStateFlow()

    /**
     * Hydrate from the audit log + [LanguagePreference]. Called once at
     * composition root; safe to call again after a `Reset device`.
     */
    suspend fun hydrate() {
        runCatching {
            auditDao.entries(AuditAction.ClinicConfigured.canonical).firstOrNull()
        }.getOrNull()?.let { entry ->
            val metadata = decode(entry.metadataJson)
            _clinicName.value = metadata["clinic_name"]
            _jurisdiction.value = metadata["jurisdiction"]
            _lawfulBasis.value = metadata["lawful_basis"]
        } ?: run {
            _clinicName.value = null
            _jurisdiction.value = null
            _lawfulBasis.value = null
        }

        runCatching {
            auditDao.entries(AuditAction.InferencePolicySet.canonical).firstOrNull()
        }.getOrNull()?.let { entry ->
            val metadata = decode(entry.metadataJson)
            metadata["threshold"]?.toDoubleOrNull()?.let { _threshold.value = it }
            metadata["default_model"]?.let { _defaultModelId.value = it }
            metadata["auto_logout_minutes"]?.toIntOrNull()?.let { _autoLogoutMinutes.value = it }
        }

        // Walk audit log for the most-recent change-of-value entries so a
        // post-onboarding edit wins over the original `inference_policy_set`.
        runCatching {
            auditDao.entries(AuditAction.ThresholdChanged.canonical).firstOrNull()
        }.getOrNull()?.let { entry ->
            decode(entry.metadataJson)["new_value"]?.toDoubleOrNull()?.let { _threshold.value = it }
        }
        runCatching {
            auditDao.entries(AuditAction.DefaultModelChanged.canonical).firstOrNull()
        }.getOrNull()?.let { entry ->
            decode(entry.metadataJson)["new_value"]?.let { _defaultModelId.value = it }
        }
        runCatching {
            auditDao.entries(AuditAction.AutoLogoutChanged.canonical).firstOrNull()
        }.getOrNull()?.let { entry ->
            decode(entry.metadataJson)["new_value"]?.toIntOrNull()?.let { _autoLogoutMinutes.value = it }
        }

        languagePreference?.let { _language.value = it.get().canonical }
    }

    // -- Edits ------------------------------------------------------------

    suspend fun updateThreshold(newValue: Double) {
        val old = _threshold.value
        _threshold.value = newValue
        val actor = currentActor()
        audit.write(
            action = AuditAction.ThresholdChanged,
            actorId = actor.id,
            actorRoleAtTime = actor.role,
            metadata = mapOf("old_value" to old.toString(), "new_value" to newValue.toString()),
        )
    }

    suspend fun updateDefaultModel(newValue: String) {
        val old = _defaultModelId.value
        _defaultModelId.value = newValue
        val actor = currentActor()
        audit.write(
            action = AuditAction.DefaultModelChanged,
            actorId = actor.id,
            actorRoleAtTime = actor.role,
            metadata = mapOf("old_value" to old, "new_value" to newValue),
        )
    }

    suspend fun updateAutoLogoutMinutes(newValue: Int) {
        val old = _autoLogoutMinutes.value
        _autoLogoutMinutes.value = newValue
        val actor = currentActor()
        audit.write(
            action = AuditAction.AutoLogoutChanged,
            actorId = actor.id,
            actorRoleAtTime = actor.role,
            metadata = mapOf("old_value" to old.toString(), "new_value" to newValue.toString()),
        )
    }

    suspend fun updateLanguage(newValue: OnboardingLanguage) {
        val old = _language.value
        _language.value = newValue.canonical
        languagePreference?.set(newValue)
        val actor = currentActor()
        audit.write(
            action = AuditAction.LanguageChanged,
            actorId = actor.id,
            actorRoleAtTime = actor.role,
            metadata = mapOf("old_value" to old, "new_value" to newValue.canonical),
        )
    }

    suspend fun updateInitials(newValue: String?) {
        val profile = runCatching { clinicians.current() }.getOrNull() ?: return
        clinicians.updateInitials(profile.id, newValue)
        audit.write(
            action = AuditAction.ProfileUpdated,
            actorId = profile.actorId,
            actorRoleAtTime = profile.role,
            resourceType = "clinician",
            resourceId = profile.actorId,
            metadata = mapOf("field" to "initials"),
        )
    }

    // -- Private ----------------------------------------------------------

    private data class Actor(val id: String, val role: String)

    private suspend fun currentActor(): Actor {
        val profile = runCatching { clinicians.current() }.getOrNull()
        return Actor(profile?.actorId ?: "unknown", profile?.role ?: "unknown")
    }

    /**
     * Parse the same flat string-map shape that `encodeMetadata` writes
     * in `AuditLog.kt`. Lenient — unknown/malformed JSON yields an empty
     * map so callers can fall back to defaults.
     */
    private fun decode(json: String): Map<String, String> {
        if (json.isBlank() || json == "{}") return emptyMap()
        // Tiny parser scoped to the shape AuditLog actually writes:
        // a flat string-to-string object with escaped string values.
        val out = mutableMapOf<String, String>()
        var i = 0
        val s = json
        if (s.isEmpty() || s[0] != '{') return emptyMap()
        i = 1
        while (i < s.length) {
            // Skip whitespace
            while (i < s.length && s[i].isWhitespace()) i++
            if (i < s.length && s[i] == '}') return out
            // Expect key
            if (i >= s.length || s[i] != '"') return out
            i++
            val keyBuilder = StringBuilder()
            while (i < s.length && s[i] != '"') {
                if (s[i] == '\\' && i + 1 < s.length) {
                    keyBuilder.append(unescape(s[i + 1]))
                    i += 2
                } else {
                    keyBuilder.append(s[i]); i++
                }
            }
            i++ // closing quote
            while (i < s.length && s[i].isWhitespace()) i++
            if (i >= s.length || s[i] != ':') return out
            i++
            while (i < s.length && s[i].isWhitespace()) i++
            if (i >= s.length || s[i] != '"') return out
            i++
            val valueBuilder = StringBuilder()
            while (i < s.length && s[i] != '"') {
                if (s[i] == '\\' && i + 1 < s.length) {
                    valueBuilder.append(unescape(s[i + 1]))
                    i += 2
                } else {
                    valueBuilder.append(s[i]); i++
                }
            }
            i++ // closing quote
            out[keyBuilder.toString()] = valueBuilder.toString()
            while (i < s.length && s[i].isWhitespace()) i++
            if (i < s.length && s[i] == ',') { i++; continue }
            if (i < s.length && s[i] == '}') return out
            return out
        }
        return out
    }

    private fun unescape(c: Char): Char = when (c) {
        'n' -> '\n'
        'r' -> '\r'
        't' -> '\t'
        'b' -> '\b'
        'f' -> ''
        '"' -> '"'
        '\\' -> '\\'
        else -> c
    }
}

File: iosApp/Services/SettingsStore.swift
import Foundation
import Observation

/// `@Observable` settings facade backed by the audit log (spec §11, §8).
///
/// Clinic-level configuration (name, jurisdiction, lawful basis) and the
/// inference policy (threshold, default model, auto-logout) live in the
/// audit log as the source of truth — they are written during Phase 1
/// admin provisioning (`clinic_configured`, `inference_policy_set`) and
/// updated by Settings edits via `threshold_changed`, `default_model_changed`,
/// `auto_logout_changed`. Holding the latest values here avoids re-parsing
/// metadata JSON on every render.
///
/// `language` lives in `UserDefaults` (`LanguagePreference`) because it
/// must survive `Reset device` (spec §10) where the audit log gets a fresh
/// `device_reprovisioned` entry but keeps every prior provisioning entry —
/// the chain-of-custody requirement.
@Observable
@MainActor
final class SettingsStore {

    // Clinic identity — read from the most recent `clinic_configured`
    // audit entry's metadata. Read-only in Settings per spec §11 (the
    // clinic config can only change via a full device reset).
    private(set) var clinicName: String?
    private(set) var jurisdiction: String?
    private(set) var lawfulBasis: String?

    // Inference policy — read from the most recent `inference_policy_set`
    // entry, then mutated through the Settings edit screens. Editable by
    // admin only per spec §11 + §9 (fresh biometric required).
    private(set) var threshold: Double = 0.3
    private(set) var defaultModelId: String = "BNLeaky_Keras"
    private(set) var autoLogoutMinutes: Int = 15

    /// Persisted UI language. Mirrors `LanguagePreference` raw value.
    private(set) var language: String = OnboardingLanguage.english.rawValue

    private let auditRepo: AuditRepository
    private let audit: AuditLog
    private let clinicians: ClinicianRepository

    init(auditRepo: AuditRepository, audit: AuditLog, clinicians: ClinicianRepository) {
        self.auditRepo = auditRepo
        self.audit = audit
        self.clinicians = clinicians
    }

    /// Hydrate from the audit log + `LanguagePreference`. Called once at
    /// composition root; safe to call again after a `Reset device`.
    func hydrate() {
        if let clinicEntry = try? auditRepo.entries(forAction: .clinicConfigured).first {
            let metadata = decode(clinicEntry.metadataJson)
            clinicName = metadata["clinic_name"]
            jurisdiction = metadata["jurisdiction"]
            lawfulBasis = metadata["lawful_basis"]
        } else {
            clinicName = nil
            jurisdiction = nil
            lawfulBasis = nil
        }

        if let policy = try? auditRepo.entries(forAction: .inferencePolicySet).first {
            let metadata = decode(policy.metadataJson)
            if let t = metadata["threshold"].flatMap(Double.init) { threshold = t }
            if let m = metadata["default_model"] { defaultModelId = m }
            if let a = metadata["auto_logout_minutes"].flatMap(Int.init) { autoLogoutMinutes = a }
        }

        // Walk audit log for the most-recent change-of-value entries so a
        // post-onboarding edit wins over the original `inference_policy_set`.
        if let last = try? auditRepo.entries(forAction: .thresholdChanged).first,
           let newValue = decode(last.metadataJson)["new_value"].flatMap(Double.init) {
            threshold = newValue
        }
        if let last = try? auditRepo.entries(forAction: .defaultModelChanged).first,
           let newValue = decode(last.metadataJson)["new_value"] {
            defaultModelId = newValue
        }
        if let last = try? auditRepo.entries(forAction: .autoLogoutChanged).first,
           let newValue = decode(last.metadataJson)["new_value"].flatMap(Int.init) {
            autoLogoutMinutes = newValue
        }

        language = LanguagePreference.get().rawValue
    }

    // MARK: - Edits (each writes the matching audit entry)

    /// Update the decision threshold. Writes a `threshold_changed` audit
    /// entry carrying old + new values per spec §11.
    func updateThreshold(_ newValue: Double) {
        let old = threshold
        threshold = newValue
        let actor = currentActor()
        audit.write(
            .thresholdChanged,
            actorId: actor.id,
            actorRoleAtTime: actor.role,
            metadata: ["old_value": String(old), "new_value": String(newValue)]
        )
    }

    func updateDefaultModel(_ newValue: String) {
        let old = defaultModelId
        defaultModelId = newValue
        let actor = currentActor()
        audit.write(
            .defaultModelChanged,
            actorId: actor.id,
            actorRoleAtTime: actor.role,
            metadata: ["old_value": old, "new_value": newValue]
        )
    }

    func updateAutoLogoutMinutes(_ newValue: Int) {
        let old = autoLogoutMinutes
        autoLogoutMinutes = newValue
        let actor = currentActor()
        audit.write(
            .autoLogoutChanged,
            actorId: actor.id,
            actorRoleAtTime: actor.role,
            metadata: ["old_value": String(old), "new_value": String(newValue)]
        )
    }

    func updateLanguage(_ newValue: OnboardingLanguage) {
        let old = language
        language = newValue.rawValue
        LanguagePreference.set(newValue)
        let actor = currentActor()
        audit.write(
            .languageChanged,
            actorId: actor.id,
            actorRoleAtTime: actor.role,
            metadata: ["old_value": old, "new_value": newValue.rawValue]
        )
    }

    /// Update microscopist initials. Writes a `profile_updated` audit entry
    /// with `field=initials` so a downstream auditor can scope queries.
    func updateInitials(_ newValue: String?) throws {
        guard let profile = try clinicians.current() else { return }
        try clinicians.updateInitials(newValue, on: profile)
        audit.write(
            .profileUpdated,
            actorId: profile.actorId,
            actorRoleAtTime: profile.role,
            resourceType: "clinician",
            resourceId: profile.actorId,
            metadata: ["field": "initials"]
        )
    }

    // MARK: - Private

    private struct Actor { let id: String; let role: String }

    private func currentActor() -> Actor {
        let profile = try? clinicians.current()
        return Actor(id: profile?.actorId ?? "unknown", role: profile?.role ?? "unknown")
    }

    private func decode(_ json: String) -> [String: String] {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        var out: [String: String] = [:]
        for (k, v) in obj {
            if let s = v as? String { out[k] = s }
            else { out[k] = "\(v)" }
        }
        return out
    }
}
